### PR TITLE
added stopwords cleaner with tests

### PR DESCRIPTION
**Pull request by shivam5992**

### CODE REVIEW

**Comment by larsga** [comment]
This is an interesting addition! For applications that match text this looks very useful. Thank you.

The build fails with a compile error: /home/travis/build/larsga/Duke/duke-core/src/main/java/no/priv/garshol/duke/cleaners/StopwordsCleaner.java:[49,21] error: cannot find symbol 

Looks like you're using a newer Java version. Please fix that.

Also, please remove the .idea folder from the PR.

I'll go through the code and comment on that now.

File: duke-core/src/main/resources/no/priv/garshol/duke/english-stopwords.txt [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+it
+itd
+it'll
+its
+itself
+i've
+j
+just
+k
+keep	keeps
```
Comment (larsga) [inline duke-core/src/main/resources/no/priv/garshol/duke/english-stopwords.txt]: A mistake here?

File: duke-core/src/test/java/no/priv/garshol/duke/cleaners/StopwordsCleanerTest.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import static junit.framework.Assert.assertEquals;
+
+public class StopwordsCleanerTest extends LowerCaseNormalizeCleanerTest {
+
+    public void setUp() {
+        cleaner = new StopwordsCleaner();
+    }
+
+    public void testMapping() {
+        assertEquals("Hello my name is duke", cleaner.clean("hello name duke"));
```
Comment (larsga) [inline duke-core/src/test/java/no/priv/garshol/duke/cleaners/StopwordsCleanerTest.java]: This test looks wrong. Surely the two string literals should be swapped?

Please also add two more tests: one with null and one with "". We need to know the cleaner can handle these two cases.

File: duke-core/src/main/java/no/priv/garshol/duke/cleaners/StopwordsCleaner.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.List;
+import java.util.ArrayList;
+
+/**
+ * A cleaner which removes english stopwords from a string.
+ */
+
+public class StopwordsCleaner implements Cleaner {
+    private LowerCaseNormalizeCleaner sub;
+    private String[] stopwords;
```
Comment (larsga) [inline duke-core/src/main/java/no/priv/garshol/duke/cleaners/StopwordsCleaner.java]: Searching this array for stop words is very inefficient. Please use a HashSet instead. That will be much faster and also easier to use.

File: duke-core/src/main/java/no/priv/garshol/duke/cleaners/StopwordsCleaner.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    private ArrayList<String> wordsList = new ArrayList<String>();
+
+
+    public StopwordsCleaner() {
+        this.sub = new LowerCaseNormalizeCleaner();
+
+        try {
+            this.stopwords = loadStopwords();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Comment (larsga) [inline duke-core/src/main/java/no/priv/garshol/duke/cleaners/StopwordsCleaner.java]: Please throw DukeException instead.

File: duke-core/src/main/java/no/priv/garshol/duke/cleaners/StopwordsCleaner.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+
+        String[] words = value.split(" ");
+        for (String word : words) {
+            wordsList.add(word);
+        }
+
+        for (int j = 0; j < stopwords.length; j++) {
+            if (wordsList.contains(stopwords[j])) {
+                wordsList.remove(stopwords[j]);
```
Comment (larsga) [inline duke-core/src/main/java/no/priv/garshol/duke/cleaners/StopwordsCleaner.java]: This is also inefficient, because `.remove` requires the rest of the array to be copied downwards. So performance will be O(n^2). I'd propose merging the two for loops. Basically, turn the top loop into

```
for (String word : words) {
  if (!stopwords.contains(word))
    wordsList.add(word);
}
```

This should be not just simpler but also much more efficient.

**Comment by shivam5992** [comment]
Okay, I will update this.

**Comment by larsga** [comment]
Thank you! That's much appreciated. I hope you agree with the comments. If not, just let me know.

**Comment by shivam5992** [comment]
Larsga, 

I have updated every point except for - 
1. .idea folder
2. the test case which you mentioned. 

Thanks
